beamer/state_machine.py

```diff
@@ -241,9 +242,18 @@ def _handle_request_resolved(event: RequestResolved, context: Context) -> Handle
     return True, None
 
 
-def _handle_fill_hash_invalidated(_event: FillHashInvalidated, _context: Context) -> HandlerResult:
-    # TODO: implement once contract is finished
-    return False, None
+def _handle_fill_hash_invalidated(event: FillHashInvalidated, context: Context) -> HandlerResult:
+    request = context.requests.get(event.request_id)
+    if request is None:
+        return False, None
+
+    # Mark the claims with that fill_id as invalidated
+    for claim in context.claims:
+        fill_hash_of_claim = request.fill_hash_with_fill_id(claim.latest_claim_made.fill_id)
+        if claim.request_id == request.id and encode_hex(fill_hash_of_claim) == event.fill_hash:
```

Comment:
Nitpick: if the claim does not belong to this request, we could avoid even computing the fill hash (not that I expect it to be too expensive).

Reply:
Fixed